contracts/Registry/TokenCuratedRegistry.sol

```diff
@@ -1,59 +1,106 @@
+import '../Challenge/IChallengeFactory.sol';
+import '../Challenge/IChallenge.sol';
 import 'openzeppelin-solidity/contracts/token/ERC20/ERC20.sol';
 
 contract TokenCuratedRegistry is StakedRegistry {
   uint public applicationPeriod;
-  ChallengeFactory public challengeFactory;
+  IChallengeFactory public challengeFactory;
 
   modifier itemNotLockedInChallenge(bytes32 id) {
     require(!inChallengePhase(id));
     _;
   }
 
+  modifier itemWhitelisted(bytes32 id) {
+    require(!inApplicationPhase(id));
+    _;
+  }
+
+  mapping(bytes32 => bytes32) applicationItems;
   mapping(bytes32 => ItemCurationData) itemsCurationData;
 
   struct ItemCurationData {
     uint applicationExpiry;
-    bool whitelisted;
-    address challengeAddress;
+    IChallenge challengeAddress;
     bool challengeResolved;
   }
 
   constructor (
     ERC20 _token,
     uint _minStake,
     uint _applicationPeriod,
-    ChallengeFactory _challengeFactory
+    IChallengeFactory _challengeFactory
   ) public
     StakedRegistry(_token, _minStake)
   {
     applicationPeriod = _applicationPeriod;
     challengeFactory = _challengeFactory;
   }
 
+  function apply(bytes32 data) public returns (bytes32) {
+    require(token.transferFrom(msg.sender, this, minStake));
+    bytes32 id = keccak256(data);
+
+    applicationItems[id] = data;
+    itemsMetadata[id] = ItemMetadata(msg.sender, minStake);
+    itemsCurationData[id] = ItemCurationData(now + applicationPeriod, IChallenge(0), false);
+  }
+
   function add(bytes32 data) public returns (bytes32) {
     bytes32 id = keccak256(data);
-    itemsCurationData[id] = ItemCurationData(now + applicationPeriod, false, address(0), false);
-    return super.add(data);
+    require(itemsCurationData[id].applicationExpiry < now);
+    require(inApplicationPhase(id) && !inChallengePhase(id));
+
+    delete applicationItems[id];
+    items[id] = data;
   }
 
-  function remove(bytes32 id) public itemNotLockedInChallenge(id) {
+  function remove(bytes32 id) public itemWhitelisted(id) itemNotLockedInChallenge(id) {
     delete itemsCurationData[id];
     super.remove(id);
   }
 
+  function resolveChallenge(bytes32 id) public {
+    IChallenge challenge = itemsCurationData[id].challengeAddress;
+    require(inChallengePhase(id));
+    require(challenge.ended());
+
+    if (challenge.passed()) {
+      _redistributeItemStake(id);
+      _rejectItem(id);
+    } else {
+      itemsCurationData[id].challengeResolved = true;
+    }
+  }
+
   function inChallengePhase(bytes32 id) public returns (bool) {
     ItemCurationData storage itemCurationData = itemsCurationData[id];
 
-    if (itemCurationData.challengeAddress == address(0)) {
+    if (itemCurationData.challengeAddress == IChallenge(0)) {
       return false;
     } else if (itemCurationData.challengeResolved) {
       return false;
     } else {
       return true;
     }
   }
+
+  function inApplicationPhase(bytes32 id) public returns (bool) {
```

Comment:
Would it be possible for this to be false even with the id still in `applicationItmes` if the application phase has expired but a function still hasn't been called which would cause the id to be removed? Or is it guaranteed somehow that the id would be removed before the application phase has expired?